Preliminary refactoring of collections API

emmileaf opened the pull request:
(For issue #667 - not ready for merge into master, but any feedback would be super helpful 🙂 )

In this PR:

- Refactored classes within the collection API, summarized by the following signatures:

```
public abstract class DocumentCollection<T extends SourceDocument> implements Iterable<FileSegment<T>>
public abstract class FileSegment<T extends SourceDocument> implements Iterable<T>
public interface SourceDocument
```

- This is a bit different than what we proposed in the issue, but I kept `DocumentCollection` and `FileSegment` as abstract classes that implement the Iterable interface since there is quite a bit of common code repeated with each implementation of a collection subclass. We can iterate over collections and segments through the `Iterable` interface (given collectionClass and collectionPath):

```
DocumentCollection collection = (DocumentCollection) collectionClass.newInstance();
testCollection.setCollectionPath(collectionPath);

for (Object s : collection){
  FileSegment segment = (FileSegment) s; 
  for (Object d : segment){
    SourceDocument doc = (SourceDocument) d; 
    foo(d);
  }
}
```
- `IndexCollection`, `MapCollections`, and tests were modified so that they are compatible with the refactored code, but major changes to the way they iterate over collections are held off for now since the current diff is pretty big with a bunch of entangled changes already.

lintool commented on src/main/java/io/anserini/collection/FileSegment.java:
```diff
@@ -0,0 +1,140 @@
+public abstract class FileSegment<T extends SourceDocument> implements Iterable<T> {
+
+  public enum Status {
+    SKIPPED, ERROR, VOID
```
does `VOID` actually mean `EOF`?

lintool commented on src/main/java/io/anserini/collection/FileSegment.java:
```diff
@@ -0,0 +1,140 @@
+   * @throws IOException if reader error encountered
+   */
+
+  protected abstract void readNext() throws IOException;
```
Since we're refactoring here, does it make sense to have `readNext()` here anymore?

emmileaf replied:
If we keep both abstract `readNext()` and abstract` createFileSegment()`, different concrete classes can share the same iterator implementations defined in `DocumentCollection` and `FileSegment`. This way, each time we add a new collection, it can just implement these two helper methods instead of overriding the entire iterator?

lintool replied:
Okay, makes sense!

lintool commented on src/main/java/io/anserini/collection/DocumentCollection.java:
```diff
@@ -88,19 +93,72 @@ public final Path getCollectionPath() {
+   * @return {@code FileSegment} with the specified path
+   * @throws IOException if file access error encountered
+   */
+  public abstract FileSegment<T> createFileSegment(Path p) throws IOException;
```
Do we still need this? Or shall we let the iterator figure out how to this itself?

lintool commented:
@emmileaf thanks, this looks great!

A few comments for you to think about... at this point you probably have a better sense of all the details...

Have you run some of the regressions to make sure things still work? (At least the small ones... we can save the large ones for when `tuna` is less busy...)

emmileaf commented:
> @emmileaf thanks, this looks great!
> 
> A few comments for you to think about... at this point you probably have a better sense of all the details...
> 
> Have you run some of the regressions to make sure things still work? (At least the small ones... we can save the large ones for when `tuna` is less busy...)

Thanks for the help! I've only had a chance to run the robust04 regression for now, but will try to run a few more later (probably after iterating over some changes) and update.

emmileaf commented:
Added some exception handling changes to the previous `FileSegment` iterator:
- Three statuses `VOID`, `SKIPPED`, and `ERROR` were used to keep track of different cases of exception handling within the iterator. `TrecwebCollection` and `TweetCollection` were throwing unchecked `RunTimeException` for skipping documents, because of limitations in the `hasNext()` signature

Changed to:
- `readNext()` now throws three types of exceptions which are caught in `hasNext()`: `IOException` (error), `NoSuchElementException` (EOF), and `ParseException` (skip). Each segment instance keeps count of whether error occurred (iteration stopped) as well as number of skips (iteration continued)

So far the following regressions pass (still working on some of the larger ones):

```
nohup python src/main/python/run_regression.py --collection disk12 --index >& log.disk12 &
nohup python src/main/python/run_regression.py --collection robust04 --index >& log.robust04 &
nohup python src/main/python/run_regression.py --collection robust05 --index >& log.robust05 &
nohup python src/main/python/run_regression.py --collection core17 --index >& log.core17 &
nohup python src/main/python/run_regression.py --collection core18 --index >& log.core18 &
nohup python src/main/python/run_regression.py --collection wt10g --index >& log.wt10g &
nohup python src/main/python/run_regression.py --collection mb11 --index >& log.mb11 &
nohup python src/main/python/run_regression.py --collection car17v1.5 --index >& log.car17v1.5 &

emmileaf commented:
I'll go ahead and merge this into the `collection-refactor` branch for now. Will create a separate PR later for merge into `master`, so that the accumulated changes won't get too messy.
